src/text/parse_result.rs

```diff
@@ -0,0 +1,286 @@
+        self.description.as_deref()
+    }
+
+    /// Consumes the `IonParseError` and returns a new one with a lifetime tied to the provided
+    /// `input`.
+    ///
+    /// This is useful in situations where the input text had to be sanitized before parsing could
+    /// be attempted. An IonParseError tied to the (locally constructed) sanitized text
+    /// cannot be returned. In such cases, the IonParseError can be replaced by one that is tied
+    /// to the original input text and which retains the original description.
+    pub fn replace_input(self, input: &str) -> IonParseError {
+        IonParseError {
+            input,
+            description: self.description,
+        }
+    }
+}
+
+/// Allows an `IonParseError` to be constructed from a `(&str, ErrorKind)` tuple, which is the
+/// data provided by core `nom` parsers if they do not match the input.
+impl<'a> From<(&'a str, ErrorKind)> for IonParseError<'a> {
+    fn from((input, _error_kind): (&'a str, ErrorKind)) -> Self {
+        IonParseError::new(input)
+    }
+}
+
+/// Allows a [nom::error::Error] to be converted into an [IonParseError] by calling `.into()`.
+impl<'a> From<nom::error::Error<&'a str>> for IonParseError<'a> {
+    fn from(nom_error: NomError<&'a str>) -> Self {
+        IonParseError::new(nom_error.input)
+    }
+}
+
+/// Allows `IonParseError` to be used as the error type in various `nom` functions.
+impl<'a> ParseError<&'a str> for IonParseError<'a> {
+    fn from_error_kind(input: &'a str, _error_kind: ErrorKind) -> Self {
+        IonParseError::new(input)
+    }
+
+    fn append(_input: &'a str, _kind: ErrorKind, other: Self) -> Self {
+        // When an error stack is being built, this method is called to give the error
+        // type an opportunity to aggregate the errors into a collection or a more descriptive
+        // message. For now, we simply allow the most recent error to take precedence.
+        other
+    }
+}
+
+/// Constructs a `nom::Err::Failure` that contains an `IonParseError` describing the problem
+/// that was encountered.
+pub(crate) fn fatal_parse_error<D: Into<String>, O>(
+    input: &str,
+    description: D,
+) -> IonParseResult<O> {
+    Err(nom::Err::Failure(IonParseError::with_description(
+        input,
+        description,
+    )))
+}
+
+/// An extension trait that allows a [std::result::Result] of any kind to be mapped to an
+/// `IonParseResult` concisely.
+pub(crate) trait OrFatalParseError<T> {
+    fn or_fatal_parse_error<'a>(self, input: &'a str, label: &str) -> IonParseResult<'a, T>;
+}
+
+/// See the documentation for [OrFatalParseError].
+impl<T, E> OrFatalParseError<T> for Result<T, E>
+where
+    E: Debug,
+{
+    fn or_fatal_parse_error<'a>(self, input: &'a str, label: &str) -> IonParseResult<'a, T> {
+        match self {
+            Ok(value) => Ok(("", value)),
+            Err(error) => fatal_parse_error(input, format!("{}: {:?}", label, error)),
+        }
+    }
```

Comment:
🗺️ Note that this method includes both the label and the error itself in the `IonParseError` it returns. Additionally, the `RawTextReader` adds location information to it in the `IonError` that it returns to the user.

All of that to say: when you see something like:
```rust
let x = do_something()
    .or_fatal_parse_error("couldn't read an int")?;
```

it looks like it doesn't tell the user much, but there's plenty of information there that isn't visible at the callsite.